File: Mar 12 138. Copy List with Random Pointer (Medium)/Solution.go
// DFS O(N) | O(N)
func copyRandomList(head *Node) *Node {
    oldToCopy := make(map[*Node]*Node)
    var dfs func (old *Node) *Node
    dfs = func (old *Node) *Node{
        // Base Cases
        if old == nil{
            return nil 
        }
        if _, found := oldToCopy[old]; found{
            return oldToCopy[old]
        }
        // Create a copy of current node
        copyNode := Node{old.Val, nil, nil}
        oldToCopy[old] = &copyNode
        // DFS clone node's next and random
        copyNode.Next = dfs(old.Next)
        copyNode.Random = dfs(old.Random)
        
        return &copyNode
    }
    
    return dfs(head)
}
// Iterative O(N) | O(N)
func copyRandomList(head *Node) *Node {
    oldToCopy := make(map[*Node]*Node)
    oldToCopy[nil] = nil
    curr := head
    // Creat all nodes but not connected
    for curr != nil{
        copyNode := Node{curr.Val, nil, nil}
        oldToCopy[curr] = &copyNode
        curr = curr.Next
    }
    curr = head
    // Make all connection through hashMap
    for curr != nil{
        copyNode := oldToCopy[curr]
        copyNode.Next = oldToCopy[curr.Next]
        copyNode.Random = oldToCopy[curr.Random]
        curr = curr.Next
    }
    
    return oldToCopy[head]
}

// Iterative Weaved List O(N) | O(1)
func copyRandomList(head *Node) *Node {
    curr := head
    // Create weaved list, no connection
    for curr != nil{
        curr, curr.Next =curr.Next, &Node{Val:curr.Val, Next:curr.Next, }
    }
    
    old := head
    // Connect random
    for old != nil{
        copyNode := old.Next
        copyNode.Random = nil
        if old.Random != nil{
            copyNode.Random = old.Random.Next
        }
        old = copyNode.Next 
    }
    
    // Connect next and Unweaved list
    var newHead *Node
    if head != nil{
        newHead = head.Next
    }
    old = head
    for old != nil{
        copyNode := old.Next
        oldNext := copyNode.Next
        copyNode.Next = nil
        if oldNext != nil{
            copyNode.Next = oldNext.Next
        }
        old.Next = oldNext
        old = old.Next
    }
    return newHead
}

File: Mar 12 138. Copy List with Random Pointer (Medium)/Solution.py
# DFS O(N) | O(N)
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        # Create a hashmap to tell us whether we clone this node
        oldToCopy = {} #Original -> Copy
        
        # Using dfs to traverse node, and make copy of it
        def dfs(node):
            # Base cases
            if node is None:
                return node
            if node in oldToCopy:
                return oldToCopy[node]
            # Create a copy of current node
            copyNode = Node(node.val)
            oldToCopy[node] = copyNode
            # DFS clone node's next and random
            copyNode.next = dfs(node.next)
            copyNode.random = dfs(node.random)
            
            return copyNode
        
        return dfs(head)

# Iterative O(N) | O(N)
class Solution:
    def copyRandomList(self, head):
        oldToCopy = {None:None}
        curr = head
        # Creat all nodes but not connect
        while curr:
            copy = Node(curr.val)
            oldToCopy[curr] = copy
            curr = curr.next
        curr = head
        # Make all connection through hashMap
        while curr:
            copy = oldToCopy[curr]
            copy.next = oldToCopy[curr.next]
            copy.random = oldToCopy[curr.random]
            curr = curr.next
            
        return oldToCopy[head]

# Iterative Weaved List O(N) | O(1)
class Solution:
    def copyRandomList(self, head):
        curr = head
        
        # Create weaved list, no connection
        while curr:
            copy = Node(curr.val, curr.next)
            curr.next = copy
            curr = copy.next
        old = head
        
        # Connect random
        while old:
            copy = old.next
            copy.random = old.random.next if old.random else None
            old = copy.next
        cloneHead = head.next if head else None
        
        # Connect next and Unweaved list
        old = head
        while old:
            copy = old.next
            nextOld = copy.next
            copy.next = nextOld.next if nextOld else None
            old.next = nextOld
            old = old.next

        return cloneHead
        
        
        
